### PR TITLE
fix(core): Return homeProject when filtering workflows by project id

### PR DESCRIPTION
**Pull request by despairblue**
## Summary

When filtering workflows by project Id we'd not return the home project if the user was not the owner of the workflow. Without that the FE shows incorrect ownerships of WFs:

image

The reason for that is that that we're filtering the workflows based on `workflow.shared.projectId` which will also filter down the `workflow.shared` list to only contain the sharings matching that projectId:

https://github.com/n8n-io/n8n/blob/146f4893e754e44344ce48a61751a604d18d5c5c/packages/cli/src/databases/repositories/workflow.repository.ts#L102-L105

Later we iterate of `workflow.shared` assuming it contains all sharings:

https://github.com/n8n-io/n8n/blob/146f4893e754e44344ce48a61751a604d18d5c5c/packages/cli/src/services/ownership.service.ts#L82-L98

In raw SQL and with a query builder this can be fixed using sub=queries, but it's not possible to have this behavior with TypeORM's repository interface. 

For credentials this was fixed 3 months ago: https://github.com/n8n-io/n8n/pull/10144

This PR implements the same fix for workflows.

image

## Related Linear tickets, Github issues, and Community forum posts

This does not address what's in the description, but what this comment mentions: 
https://linear.app/n8n/issue/PAY-2271/community-issue-credentials-in-project-personal-are-not-just-my#comment-9fe3bc10

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [x] ~[Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.~
- [x] Tests included. 
- [x] ~PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)~

### CODE REVIEW

File: packages/cli/src/workflows/workflow.service.ts [hunk 1]
```diff
@@ -66,6 +66,21 @@ export class WorkflowService {
...
+			if (typeof options?.filter?.projectId === 'string' && options.filter.projectId !== '') {
+				const relations = await this.sharedWorkflowRepository.getAllRelationsForWorkflows(
+					workflows.map((c) => c.id),
+				);
+				workflows.forEach((c) => {
+					c.shared = relations.filter((r) => r.workflowId === c.id);
+				});
+			}
+
+			console.log(workflows.length);
```
Comment (r00gm) [inline packages/cli/src/workflows/workflow.service.ts:83]: missed this one :)

File: packages/cli/src/workflows/workflow.service.ts [hunk 2]
```diff
@@ -66,6 +66,20 @@ export class WorkflowService {
...
+			// To solve this, we have to get all the relation now, even though
+			// we're deleting them later.
+			if (typeof options?.filter?.projectId === 'string' && options.filter.projectId !== '') {
+				const relations = await this.sharedWorkflowRepository.getAllRelationsForWorkflows(
+					workflows.map((c) => c.id),
+				);
+				workflows.forEach((c) => {
+					c.shared = relations.filter((r) => r.workflowId === c.id);
+				});
+			}
```
Comment (ivov) [inline packages/cli/src/workflows/workflow.service.ts:81]: The facts that we're having to do two queries for this, in-memory processing and fetching things that we later discard makes me wish we could use the query builder, but that'd be a larger refactor. Maybe to improve in future.

**Review by ivov — APPROVED** [review]
Confirmed working, thanks for fixing 🙏🏻

**Comment by janober** [comment]
Got released with [`n8n@1.72.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.72.0)

**Pull request merged**